src/ngx_http_lua_socket_tcp.c

```diff
@@ -2927,6 +3086,83 @@ ngx_http_lua_socket_tcp_send_retval_handler(ngx_http_request_t *r,
     return 1;
 }
 
+static int
+ngx_http_lua_socket_tcp_select_retval_handler(ngx_http_request_t *r,
+    ngx_http_lua_socket_tcp_upstream_t *u, lua_State *L)
+{
+    int                          matched_index;
+    ngx_http_lua_ctx_t          *ctx;
+    ngx_event_t                 *ev;
+    unsigned int                            i;
+    ngx_http_lua_socket_tcp_upstream_t    *u_entry, **u_array_entries;
+    ngx_http_lua_socket_tcp_select_context_t *select_ctx;
+    ngx_http_lua_co_ctx_t                    *coctx;
+
+    ngx_http_lua_loc_conf_t             *llcf;
+
+    ngx_log_debug0(NGX_LOG_DEBUG_HTTP, r->connection->log, 0,
+                   "lua tcp socket select return value handler");
+
+    ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module);
+
+    coctx = ctx->cur_co_ctx;
+
+    select_ctx = coctx->data;
+
+#if 1
+    if (u->raw_downstream || u->body_downstream) {
+        llcf = ngx_http_get_module_loc_conf(r, ngx_http_lua_module);
+
+        if (llcf->check_client_abort) {
+
+            r->read_event_handler = ngx_http_lua_rd_check_broken_connection;
+
+            ev = r->connection->read;
+
+            dd("rev active: %d", ev->active);
+
+            if ((ngx_event_flags & NGX_USE_LEVEL_EVENT) && !ev->active) {
+                if (ngx_add_event(ev, NGX_READ_EVENT, 0) != NGX_OK) {
+                    lua_pushnil(L);
+                    lua_pushliteral(L, "failed to add event");
+                    return 2;
+                }
+            }
+
+        } else {
+            /* llcf->check_client_abort == 0 */
+            r->read_event_handler = ngx_http_block_reading;
+        }
+    }
+#endif
+
+    matched_index = -1;
+
+    u_array_entries = select_ctx->u_array.elts;
+
+    for (i = 0; i < select_ctx->u_array.nelts; i++) {
+        u_entry = u_array_entries[i];
+        if (u_entry == u) {
+            matched_index = i;
+            break;
+        }
+    }
+
+    if (u->ft_type) {
+        if (u->ft_type & NGX_HTTP_LUA_SOCKET_FT_TIMEOUT) {
+            u->no_close = 1;
+        }
+
+        lua_pushnumber(L, matched_index + 1);
+        (void) ngx_http_lua_socket_read_error_retval_handler(r, u, L);
+        lua_remove(L, -2);
+        return 2;
+    }
+
+    lua_pushnumber(L, matched_index + 1);
+    return 1;
```

Comment:
I am wonder if returning only one readable socket is OK... Please correct me if I am wrong:
At least with epoll event module, the read event is triggered only once at the edge (search EPOLLET as the keyword). If multiple events happen at the same time, and only one event is handled, the other shots will be lost.

Reply:
You're definitely right. I'm not sure how conceptually I could do that (combine all the events from the same epoll into 1 return code) but I'll look into that.
I believe that even if a solution of returning multiple FDs doesn't work out - this select() API is very relevant, as you can call select() again after you just called it and get the remaining FD(s).
I'll look into it and update.